src/storage/txn/commands/txn_heart_beat.rs

```diff
@@ -56,53 +56,26 @@ impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for TxnHeartBeat {
             .into()
         ));
 
-        let lock: MvccResult<_> = if let Some(mut lock) = txn.reader.load_lock(&self.primary_key)? {
-            if lock.ts == self.start_ts {
+        let lock = match txn.reader.load_lock(&self.primary_key)? {
+            Some(mut lock) if lock.ts == self.start_ts => {
                 if lock.ttl < self.advise_ttl {
                     lock.ttl = self.advise_ttl;
                     txn.put_lock(self.primary_key.clone(), &lock);
-                } else {
-                    debug!(
-                        "txn_heart_beat with advise_ttl not larger than current ttl";
-                        "primary_key" => %self.primary_key,
-                        "start_ts" => self.start_ts,
-                        "advise_ttl" => self.advise_ttl,
-                        "current_ttl" => lock.ttl,
-                    );
                 }
-                Ok(lock)
-            } else {
-                debug!(
-                    "txn_heart_beat invoked but lock is absent";
-                    "primary_key" => %self.primary_key,
-                    "start_ts" => self.start_ts,
-                    "advise_ttl" => self.advise_ttl,
-                );
-                Err(MvccErrorInner::TxnLockNotFound {
-                    start_ts: self.start_ts,
-                    commit_ts: TimeStamp::zero(),
-                    key: self.primary_key.clone().into_raw()?,
-                }
-                .into())
+                lock
             }
-        } else {
-            debug!(
-                "txn_heart_beat invoked but lock is absent";
-                "primary_key" => %self.primary_key,
-                "start_ts" => self.start_ts,
-                "advise_ttl" => self.advise_ttl,
-            );
-            Err(MvccErrorInner::TxnLockNotFound {
-                start_ts: self.start_ts,
-                commit_ts: TimeStamp::zero(),
-                key: self.primary_key.clone().into_raw()?,
+            _ => {
+                return Err(MvccError::from(MvccErrorInner::TxnNotFound {
```

Comment:
Looks like this skips the statistic calculation at L76 . Is it ok?

Reply:
I didn't notice it, but I find we skip this step in many places if there is an error.

Reply:
Looking like it is easy to forget adding these statistics. What about passing statistics to `MvccTxn` and adding the `MvccReader` statistics in drop? Or using some way to let `MvccReader` modify the outer statistics directly?

Reply:
Good idea. I will have a try.